src/phylo/phylogeny.jl

```diff
@@ -0,0 +1,260 @@
+
+Type parameter `C` dictates what datatype can be stored in the phylogeny to annotate
+clades and tips. Type parameter `B` dictates what datatype can be stored in the
+phylogeny to annotate branches. Think `C` for clades and `B` for branches.
+
+"""
+type Phylogeny{C, B}
+    graph::DiGraph
+    vertexindex::Indexer{Int}
+    vertexdata::Vector{C}
+    edgedata::Dict{Edge, B}
+    ntaxa::Int
+    rooted::Bool
+    rerootable::Bool
+end
+
+
+# Phylogeny Constructors
+# -----------------------
+
+"""
+    Phylogeny{C, B}(tips::Vector{Symbol}, ::Type{C}, ::Type{B})
+
+Construct a completely unresolved phylogeny, with no annotated data.
+
+The phylogeny will store annotation data of type `C` for tree nodes
+(clades and tips). It will also store data of type `B` associated with branches.
+
+# Examples
+```julia
+Phylogeny([:Human, :Chimp, :Dog, :Cat], Void, Void)
+Phylogeny([:Human, :Chimp, :Dog, :Cat], Float64, Float64)
+```
+"""
+function Phylogeny{C, B}(::Type{C}, ::Type{B}, taxa::Vector{Symbol})
+    ntaxa = length(taxa)
+    maxVertices = maxvertices(ntaxa)
+    taxaAndRoot = push!(copy(taxa), :Root)
+    idxer = Indexer(taxaAndRoot, Int)
+    g = DiGraph(maxVertices)
+    for i in 1:ntaxa
+        add_edge!(g, ntaxa + 1, i)
+    end
+    v = Vector{C}()
+    d = Dict{Edge, B}()
+    return Phylogeny{C, B}(g, idxer, v, d, ntaxa, false, true)
+end
+
+
+"""
+    Phylogeny(taxa::Vector{Symbol})
+
+Construct a completely unresolved phylogeny, with no annotated data.
+
+This constructor assumes you want a phylogeny where the only data associated
+with it is branch lengths (Float64) for the edges, and confidence values
+(Float64) for the clades.
+
+# Examples
+```julia
+Phylogeny([:Human, :Chimp, :Dog, :Cat])
+```
+"""
+function Phylogeny(taxa::Vector{Symbol})
+    return Phylogeny(Float64, Float64, taxa)
+end
+
+
+"""
+    Phylogeny{T <: AbstractString}(taxa::Vector{T})
+Construct a completely unresolved phylogeny, with no annotated data.
+
+This constructor assumes you want a phylogeny where the only data associated
+with it is branch lengths (Float64) for the edges, and confidence values
+(Float64) for the clades.
+"""
+function Phylogeny{T <: AbstractString}(taxa::Vector{T})
+    return Phylogeny(convert(Vector{Symbol}, taxa))
+end
+
+
+function Base.copy{C, B}(tree::Phylogeny{C, B})
+    g = copy(tree.graph)
+    i = copy(tree.vertexindex)
+    v = copy(tree.vertexdata)
+    d = copy(tree.edgedata)
+    return Phylogeny{C, B}(g, i, v, d, tree.ntaxa, tree.rooted, tree.rerootable)
+end
+
+
+# Miscellaneous Functions
+# ------------------------
+
+"""
+    n_possible_rooted(n::Int)
+
+Returns the number of possible rooted tree topoligies for n taxa.
+
+**Warning:** This method currently overflows when n > 12.
+"""
+function n_possible_rooted(n::Int)
+    @assert n >= 2 "n must be greater or equal to 2."
+    num = factorial(2 * n - 3)
+    den = (2^(n - 2)) * factorial(n - 2)
+    return Int(num / den)
+end
```

Comment:
You can compute `(2n-3)!!` as `doublefactorial(2n-3)` using the [Combinatorics.jl](https://github.com/JuliaLang/Combinatorics.jl) package.